job/deploy_manager.go

```diff
@@ -0,0 +1,170 @@
+package job
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/odpf/salt/log"
+
+	"github.com/odpf/optimus/models"
+	"github.com/odpf/optimus/store"
+	"github.com/odpf/optimus/utils"
+)
+
+type DeployManager interface {
+	Init()
+	Deploy(ctx context.Context, projectSpec models.ProjectSpec) (models.DeploymentID, error)
+	GetStatus(ctx context.Context, deployID models.DeploymentID) (models.JobDeployment, error)
+}
+
+type deployManager struct {
+	l      log.Logger
+	config DeployManagerConfig
+
+	deployer         Deployer
+	deployerCapacity int32
+	uuidProvider     utils.UUIDProvider
+	deployRepository store.JobDeploymentRepository
+
+	// wait group to synchronise on deployers
+	wg sync.WaitGroup
+
+	// request queue, used by deployers
+	requestQ chan models.JobDeployment
+}
+
+type DeployManagerConfig struct {
+	NumWorkers    int
+	WorkerTimeout time.Duration
+	QueueCapacity int
+}
+
+func (m *deployManager) Deploy(ctx context.Context, projectSpec models.ProjectSpec) (models.DeploymentID, error) {
+	// Check deployment status for the requested Project
+	jobDeployment, err := m.deployRepository.GetByStatusAndProjectID(ctx, models.JobDeploymentStatusInQueue, projectSpec.ID)
+	if err == nil {
+		return jobDeployment.ID, nil
+	}
+
+	// Return valid errors
+	if !errors.Is(err, store.ErrResourceNotFound) {
```

Comment:
can you elaborate this scenario

Reply:
this means, we are not returning the error if it is ErrResourceNotFound (when no job deployment with same project spec found in queue), and return the other errors.